preprocess.py

```diff
@@ -0,0 +1,176 @@
+    return df
+
+
+def find_honorifics(name: str) -> str:
+    honorific_names = r'(?:Mrs|Mr|Ms|Miss|Master|Don|Rev|Mme|Major|Mlle|Col|Capt|Jonkheer|Countess|Dr)\.?'
+    honorific = re.findall(honorific_names, name)
+    if not honorific:  # empty list - no honorific found
+        return 'non_honorific'
+    honorific = honorific[0]  # findall returns a list - in our case list of single string
+    honorific = honorific.split()[0]  # the string contains the honorific in first place if exists
+    honorific = honorific.replace('.', '')
+    return honorific
+
+
+def honorifics_feature_from_name(df: pd.DataFrame) -> pd.DataFrame:
+    df['Honorifics'] = df['Name'].apply(lambda name: find_honorifics(name))
+    return df
+
+
+def solo_feature_from_parch_sibsp(df, y=None) -> pd.DataFrame:
+    # If both SibSp and Parch are 0 then we say the passenger is by himself on board, meaning he is traveling solo
+    df['Solo'] = (df['Parch'] + df['SibSp']).apply(lambda x: 1 if x == 0 else 0)
+    return df
+
+
+def CabinChar_feature_from_cabin(df: pd.DataFrame) -> pd.DataFrame:
+    # Extracting first character from the cabin (which is with about 70% of missing values in train)
+    cabins = df['Cabin'].astype(str)
+    cabin_char = [cab[0] for cab in cabins]  # when NaN then char would be 'n'
+    df['CabinChar'] = cabin_char
+    return df
+
+
+def TicketGroup_GroupSize_features_from_ticket(df: pd.DataFrame) -> pd.DataFrame:
+    ticket_counts_dict = df['Ticket'].value_counts().to_dict()
+    df['TicketGroup'] = df['Ticket'].apply(lambda x: 1 if ticket_counts_dict[x] > 1 else 0)
+    df['GroupSize'] = df['Ticket'].map(ticket_counts_dict)
+    return df
+
+
+def label_encoding_transformer(label_encoder_features):
+    label_encoder_transformer = OrdinalEncoder()
+    return "label_encoder", label_encoder_transformer, label_encoder_features
+
+
+def one_hot_encoding_transformer(one_hot_features):
+    one_hot_transformer = OneHotEncoder(handle_unknown="ignore", sparse_output=False)
+    return "onehot", one_hot_transformer, one_hot_features
+
+
+def numeric_imputing_transformer(numeric_features):
+    numeric_transformer = IterativeImputer(estimator=LinearRegression(), missing_values=np.nan, max_iter=10,
+                                           imputation_order='roman', random_state=0)
+    return "numeric", numeric_transformer, numeric_features
+
+
+class Preprocess:
+    def __init__(self, df: pd.DataFrame = None, df_path: str = None):
+        """
+        Will make a preprocessor sklearn ColumnTransformer for using in pipeline
+        Args:
+            df_path: path of titanic dataframe kind data with columns:
+            ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age',
+            'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']
+            df: if given otherwise read from path
+        """
+        assert not all(v is None for v in [df_path, df]), "One argument should be not None"
+        if df_path is not None:
```

Comment:
Every class should do one thing, so here the class both reads the data and preprocess it
I recommend reading the data outside of the class and processing it in the class